File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int res=nums[0],mn=res,mx=res;
        for(int i=1;i<nums.length;i++){
            if(nums[i]<0){
                int temp=mn;
                mn=mx;
                mx=temp;
            }
            mn=Math.min(nums[i],nums[i]*mn);
            mx=Math.max(nums[i],nums[i]*mx);
            res=Math.max(res,mx);
        }
        return res;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
//         int res=nums[0],mx=res,mn=res;
//         for(int i=1;i<nums.size();i++){
//             if(nums[i]<0)swap(mx,mn);
            
//             mx=max(nums[i],mx*nums[i]);
//             mn=min(nums[i],mn*nums[i]);
            
//             res=max(res,mx);
//         }
//         return res;
        
        // int ans=nums[0];
        // int mx=ans,mn=ans;
        // for(int i=1;i<nums.size();i++){
        //     if(nums[i]<0)swap(mx,mn);
        //     mx=max(nums[i],mx*nums[i]);
        //     mn=min(nums[i],mn*nums[i]);
        //     ans=max(ans,mx);
        // }
        // return ans;
        
        
        int ans=nums[0];
        int mn=ans,mx=ans;
        for(int i=1;i<nums.size();i++){
            if(nums[i]<0)swap(mn,mx);
            mx=max(mx*nums[i],nums[i]);
            mn=min(mn*nums[i],nums[i]);
            ans=max(mx,ans);
        }
        return ans;
    }
};
